### PR TITLE
Add more license annotations

### PR DESCRIPTION
**Pull request by emilyalbini**
This PR updates the `.reuse/dep5` file to include more accurate licensing data for everything in the repository (*excluding* submodules and dependencies). Some decisions were made in this PR:

* The standard copyright attribution for files maintained by us is "The Rust Project Developers (see https://thanks.rust-lang.org)", to avoid having to maintain an in-tree `AUTHORS` file.
* For files that have specific licensing terms, we added the terms to the `.reuse/dep5` rather than adding SPDX comments in the files themselves.
* REUSE picks up any comment/text line with `Copyright` on it, so I had to sprinkle around `REUSE-IgnoreStart` and `REUSE-IgnoreEnd` comments.

The rendered `COPYRIGHT` file is available at https://gist.github.com/pietroalbini/efb81103f69596d39758114f3f6a8688.

r? @pnkfelix

### CODE REVIEW

**Comment by WaffleLapkin** [comment]
@rustbot author
(marking as waiting author since this is a draft)

**Comment by pietroalbini** [comment]
Rebased and applied the recommendations from the foundation lawyer (cc @joelmarcey). Here is the newly rendered [COPYRIGHT.md](https://gist.github.com/pietroalbini/efb81103f69596d39758114f3f6a8688).

@rustbot review

**Comment by pnkfelix** [comment]
I reviewed this and was not clear about the policy/plan for the blocks surrounded by REUSE-IgnoreStart/REUSE-IgnoreEnd, so I asked @pietroalbini about it [on Zulip over here](https://rust-lang.zulipchat.com/#narrow/stream/231349-t-core.2Flicensing/topic/reuse.20ignore.20.23104527/near/340626603).

Executive Summary: The ignore directives are used to stop the REUSE tool from using the ignore-annotated copyright statements in question when building its resulting report. (as in, "it makes the tool ignore these blocks"). There are two scenarios where it is "okay" to ignore such statements:
 * Old copyright statements that don't match our current templates. 
   *  These should be ported forward to use the new standard copyright statement that is understood by the REUSE tool.
 * Text that is itself not a copyright statement (at least not about the file in question), but matches the REUSE templates and thus would be picked up without an ignore directive. @pietroalbini gives the example of `src/doc/footer.inc`, which includes text that is is auto-injected into the generated documentation in order to establish the copyright *of* the generated documentation, but that text is not itself the copyright assignment for `src/doc/footer.inc`.
   * These should either be "tweaked" to "appease" REUSE (which I take to mean "rewrite the text/code such that REUSE stops pattern matching these cases") or should keep the ignore-directives. (I think the specific example of `src/doc/footer.inc` sounds like a scenario where keeping the ignore-directive makes sense. We probably need to consider each such instance on a case-by-case basis, rather than applying a blanket policy for the long-term.)

I also asked Pietro whether our goal was to remove licenses entirely, or to replace licenses with spdx headers. Pietro said that they'd love in the long-term to have spdx headers in every file, but for the short term the focus is to minimal number of changes to get our repository into compliance (and I agree with that.) Overall, these spdx headers seem to require on the order of two lines of machine-parsed comments, so the aforementioned long-term plan seems reasonable to me

File: src/tools/miri/tests/pass/intrinsics-integer.rs [hunk 1]
```diff
@@ -1,12 +1,5 @@
-// Copyright 2012-2014 The Rust Project Developers. See the COPYRIGHT
-// file at the top-level directory of this distribution and at
-// http://rust-lang.org/COPYRIGHT.
-//
-// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
-// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
-// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
-// option. This file may not be copied, modified, or distributed
-// except according to those terms.
```
Comment (RalfJung) [inline src/tools/miri/tests/pass/intrinsics-integer.rs:9]: FWIW these are copies of old rust source files, that had their license headers removed in rustc at some point but that change was never synced to Miri. So now this 'fork' has the SPDX header but the original file that still exists in rustc does not... it'd be more consistent to just remove the license header here without adding SPDX headers. But either way is fine for me.

**Pull request merged**